### PR TITLE
Add support for comment and label manifests.

### LINKED ISSUE #1286: Design to keep PullRequest PipelineResource in sync with reality
# Expected Behavior

If a Task uses a PullRequest Pipeline resource, as the Task progresses, if comments are added (or other changes represented in the filesystem) to the Pull Request, it should be possible:

- For the executing Task to see these changes <-- this is kind of a stretch goal really
- For the executing Task to use the PullRequest resource as an output and make _additive_ changes to the PullRequest

# Actual Behavior

When a Task is executing, the representation on the file system of the pull request is _not_ updated. Even worse, the side effect is that when the PullRequestResource is used as an output, the data is treated as the source of truth and so any changes made between when the Task started and when it completed can be lost (e.g. comments made in the meantime will be deleted!!!)

# Steps to Reproduce the Problem

1. Create a Task that uses [a PullRequest Resource](https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#pull-request-resource) as an output, e.g.:
  a. Task could sleep for a minute or something (to give you time for step 3)
  b. Task could also add a comment to the output (i.e. write to /workspace/<resource>/comments/<comment>)
2. Execute the Task, providing an existing PR as input
3. Add a comment to the PR manually while the Task executes!
4. After the task finishes, observe that the PR has the comment added by the Task, but the one you added manually has vanished :(

# Additional Info

We should at _least_ make it so changes are additive, but also we could be ambitious and do something really cool e.g. create our own fs? (@dlorenc had some cool ideas here)

### PR DESCRIPTION
**Pull request by wlynch**
# Changes

Introduces `.MANIFEST` files to comment and label directories to keep
track of sub-resources when the PullRequest resource is initialized.
This allows us to keep track of what resources the user explicitly
deleted, and not accidentally overwrite new comments/lables that were
introduced during execution.

Minor changes included:
* Changed fake GitHub server to make copies of PullRequests to avoid
  pointer collision across tests.
* Changed uploadLabels to Add/Delete labels explicitly instead of
  replacing all.

Change manifest path to .MANIFEST to avoid collisions.

Fixes #1286 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
- Fixes bug where PullRequestResource could accidentally delete newly created upsteam resources in certain race conditions.
```

### CODE REVIEW

**Comment by wlynch** [comment]
/hold

Found some issues with the manifest, so this is not ready for review yet. :sweat_smile:

File: cmd/pullrequest-init/disk_test.go [hunk 1]
```diff
@@ -113,26 +115,39 @@ func TestToDisk(t *testing.T) {
...
 		}
 	}
+	gotManifest := readManifest(t, filepath.Join(d, "labels", manifestPath))
+	for _, m := range gotManifest {
+		if _, ok := labelManifest[m]; !ok {
+			t.Errorf("Label %s not found in manifest", m)
+		}
+	}
+	if len(labelManifest) != len(gotManifest) {
+		t.Errorf("Label manifest does not match expected length. %d !=  %d, got %+v", len(labelManifest), len(gotManifest), gotManifest)
```
Comment (sbwsg) [inline cmd/pullrequest-init/disk_test.go]: suggest "expected %d, received %d", just to make it clear which number was the assertion and which was the result.

File: cmd/pullrequest-init/fake_github.go [hunk 2]
```diff
@@ -251,6 +257,24 @@ func (g *FakeGitHub) updateLabels(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusOK)
 }
 
+func (g *FakeGitHub) listLabels(w http.ResponseWriter, r *http.Request) {
+	key, err := prKey(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (sbwsg) [inline cmd/pullrequest-init/fake_github.go]: return after this?

File: cmd/pullrequest-init/disk.go [hunk 3]
```diff
@@ -127,57 +159,86 @@ func refToDisk(name, path string, r *GitReference) error {
...
 	commentsPath := filepath.Join(path, "comments")
 	statusesPath := filepath.Join(path, "status")
 
 	pr := PullRequest{}
+	manifests := make(map[string]Manifest)
 
 	var err error
+	var manifest Manifest
 
 	// Start with comments
```
[the remark this answers is not in the document]
Comment (wlynch) [reply cmd/pullrequest-init/disk.go]: Done.

File: cmd/pullrequest-init/disk_test.go (hunk 1 above)
Comment (wlynch) [reply cmd/pullrequest-init/disk_test.go]: Done.

File: cmd/pullrequest-init/fake_github.go (hunk 2 above)
Comment (wlynch) [reply cmd/pullrequest-init/fake_github.go]: Done.

**Comment by wlynch** [comment]
Squashed the commits in prep for submission.

File: cmd/pullrequest-init/disk.go [hunk 4]
```diff
@@ -80,28 +86,54 @@ func ToDisk(pr *PullRequest, path string) error {
...
+	f, err := os.Create(filepath.Join(path, manifestPath))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	manifest := bufio.NewWriter(f)
+
 	for _, c := range comments {
-		commentPath := filepath.Join(path, strconv.FormatInt(c.ID, 10)+".json")
+		id := strconv.FormatInt(c.ID, 10)
```
Comment (dlorenc) [inline cmd/pullrequest-init/disk.go]: Is this manifest intended to be human readable? Would json serializing or gob serializing make it easier to safely round trip, so we don't have to manually encode to text?

Comment (wlynch) [reply cmd/pullrequest-init/disk.go]: Great idea!

Manifests are not meant to be human readable. Switched to gob encoding.

**Pull request merged**